File: common/src/main/java/cc/turtl/chiselmon/api/filter/FilterTagParser.java
package cc.turtl.chiselmon.api.filter;

import cc.turtl.chiselmon.api.predicate.PokemonPredicates;
import cc.turtl.chiselmon.util.ParseUtils;
import com.cobblemon.mod.common.pokemon.Gender;
import com.cobblemon.mod.common.pokemon.Pokemon;

import java.util.function.Predicate;
import java.util.stream.StreamSupport;

public class FilterTagParser {

    /**
     * Parses a simple tag string into a Pokemon predicate.
     * <p>
     * Examples:
     * - "shiny" -> IS_SHINY
     * - "legendary" -> IS_LEGENDARY
     * - "type=fire" -> has fire type
     * - "species=pikachu" -> is Pikachu
     * - "min_size=1.5" -> sizes inclusive above 1.5
     * - "max_size=0.4" -> sizes inclusive below 0.4
     * - "gender=male" -> is male
     */
    public static Predicate<Pokemon> parse(String tag) {
        tag = tag.toLowerCase().trim();

        return switch (tag) {
            case "shiny" -> PokemonPredicates.IS_SHINY;
            case "legendary" -> PokemonPredicates.IS_LEGENDARY;
            case "extreme_size" -> PokemonPredicates.IS_EXTREME_SIZE;
            default -> parseComplexTag(tag);
        };
    }

    private static Predicate<Pokemon> parseComplexTag(String tag) {
        if (!tag.contains("=")) {
            return p -> false;
        }

        String[] parts = tag.split("=", 2);
        String key = parts[0].toLowerCase().trim();
        String rawValue = parts[1].trim();

        return switch (key) {
            case "species" -> {
                String normalizedSearch = ParseUtils.normalizeSpeciesName(rawValue);
                yield pokemon -> {
                    String internalName = ParseUtils.normalizeSpeciesName(pokemon.getSpecies().getName());
                    return internalName.equals(normalizedSearch);
                };
            }

            case "type" -> pokemon -> StreamSupport.stream(pokemon.getTypes().spliterator(), false)
                    .anyMatch(t -> t.getName().equalsIgnoreCase(rawValue));

            case "gender" -> pokemon -> {
                try {
                    return pokemon.getGender().equals(Gender.valueOf(rawValue.toUpperCase()));
                } catch (Exception e) {
                    return false;
                }
            };

            case "min_size" -> pokemon -> {
                try {
                    return pokemon.getScaleModifier() >= Float.parseFloat(rawValue);
                } catch (Exception e) {
                    return false;
                }
            };

            case "max_size" -> pokemon -> {
                try {
                    return pokemon.getScaleModifier() <= Float.parseFloat(rawValue);
                } catch (Exception e) {
                    return false;
                }
            };

            case "min_level" -> pokemon -> {
                try { return pokemon.getLevel() >= Integer.parseInt(rawValue); }
                catch (Exception e) { return false; }
            };

            default -> p -> false;
        };
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/api/filter/FilterTagParser.kt
package cc.turtl.chiselmon.core.api.filter

import cc.turtl.chiselmon.core.api.predicate.IS_EXTREME_SIZE
import cc.turtl.chiselmon.core.api.predicate.IS_LEGENDARY
import cc.turtl.chiselmon.core.api.predicate.IS_SHINY
import cc.turtl.chiselmon.core.util.normalizeSpeciesName
import com.cobblemon.mod.common.pokemon.Gender
import com.cobblemon.mod.common.pokemon.Pokemon
import java.util.function.Predicate

/**
 * Parses a simple tag string into a [Predicate] for [Pokemon].
 *
 * Supported tags:
 * - "shiny" -- matches shiny pokemon
 * - "legendary" -- matches legendary pokemon
 * - "extreme_size" -- matches pokemon with extreme size
 * - "type=fire" -- matches pokemon with the given type
 * - "species=pikachu" -- matches a specific species
 * - "form=alolan" -- matches alolan forms
 * - "gender=male" -- matches a specific gender
 * - "min_size=1.5" -- matches pokemon at or above the given scale
 * - "max_size=0.4" -- matches pokemon at or below the given scale
 * - "min_level=50" -- matches pokemon at or above the given level
 */
object FilterTagParser {

    @JvmStatic
    fun parse(tag: String): Predicate<Pokemon> {
        return when (val normalized = tag.lowercase().trim()) {
            "shiny" -> IS_SHINY
            "legendary" -> IS_LEGENDARY
            "extreme_size" -> IS_EXTREME_SIZE
            else -> parseComplexTag(normalized)
        }
    }

    private fun parseComplexTag(tag: String): Predicate<Pokemon> {
        if (!tag.contains("=")) return Predicate { false }

        val (key, rawValue) = tag.split("=", limit = 2).let {
            it[0].trim() to it[1].trim()
        }

        return when (key) {
            "species" -> {
                val normalizedSearch = normalizeSpeciesName(rawValue)
                Predicate { pokemon ->
                    normalizeSpeciesName(pokemon.species.name) == normalizedSearch
                }
            }

            "form" -> Predicate { pokemon ->
                pokemon.form.name.equals(rawValue, ignoreCase = true)
            }

            "type" -> Predicate { pokemon ->
                pokemon.types.any { it.name.equals(rawValue, ignoreCase = true) }
            }

            "gender" -> {
                val gender = try {
                    Gender.valueOf(rawValue.uppercase())
                } catch (_: IllegalArgumentException) {
                    throw FilterConditionParser.ParseException(
                        "Invalid gender '$rawValue' -- expected male, female, or genderless"
                    )
                }
                Predicate { it.gender == gender }
            }

            "min_size" -> {
                val threshold = rawValue.toFloatOrNull()
                    ?: throw FilterConditionParser.ParseException("Invalid min_size value: '$rawValue'")
                Predicate { it.scaleModifier >= threshold }
            }

            "max_size" -> {
                val threshold = rawValue.toFloatOrNull()
                    ?: throw FilterConditionParser.ParseException("Invalid max_size value: '$rawValue'")
                Predicate { it.scaleModifier <= threshold }
            }

            "min_level" -> {
                val level = rawValue.toIntOrNull()
                    ?: throw FilterConditionParser.ParseException("Invalid min_level value: '$rawValue'")
                Predicate { it.level >= level }
            }

            else -> Predicate { false }
        }
    }
}
